tests/symbols-used.rs

```diff
@@ -0,0 +1,99 @@
+#![cfg(feature = "internal")]
+
+// This test checks that all symbols defined in Clippy's `sym.rs` file
+// are used in Clippy. Otherwise, it will fail with a list of symbols
+// which are unused.
+//
+// This test only runs when the `internal` feature is set, to allow
+// developers to temporarily define new, yet unused symbols without
+// failing the tests. Also, the test is a no-op if run as part of the
+// compiler test suite.
+
+use std::collections::HashSet;
+use std::path::Path;
+
+use futures::StreamExt;
+use futures::stream::FuturesUnordered;
+use regex::Regex;
+use walkdir::{DirEntry, WalkDir};
+
+const SYM_FILE: &str = "clippy_utils/src/sym.rs";
+
+type Result<T, E = AnyError> = std::result::Result<T, E>;
+type AnyError = Box<dyn std::error::Error>;
+
+async fn load_interned_symbols() -> Result<HashSet<String>> {
+    let content = tokio::fs::read_to_string(SYM_FILE).await?;
+    let content = content
+        .split_once("generate! {")
+        .ok_or("cannot find symbols start")?
+        .1
+        .split_once("\n}\n")
+        .ok_or("cannot find symbols end")?
+        .0;
+    let re = Regex::new(r"(?m)^    (\w+)").unwrap();
+    Ok(re.captures_iter(content).map(|m| m[1].to_owned()).collect())
+}
+
+async fn load_symbols(file: impl AsRef<Path>, re: &Regex) -> Result<Vec<String>> {
+    let content = tokio::fs::read_to_string(file).await?;
+    Ok(re.captures_iter(&content).map(|m| m[1].to_owned()).collect())
+}
+
+async fn load_paths(file: impl AsRef<Path>, re: &Regex) -> Result<Vec<String>> {
+    let content = tokio::fs::read_to_string(file).await?;
+    Ok(re
+        .captures_iter(&content)
+        .flat_map(|m| m[1].split("::").map(String::from).collect::<Vec<_>>())
+        .collect())
+}
+
+#[tokio::test]
+#[allow(clippy::case_sensitive_file_extension_comparisons)]
+async fn all_symbols_are_used() -> Result<()> {
```

Comment:
Since we're just using tokio for its threadpool here perhaps `used_symbols` should use rayon instead, the other things can run serially

Reply:
Ok, I was reluctant to add `rayon` as a new dependency, I'll make the change.

Reply:
Done

Reply:
Oh right it's a lintcheck dependency, I thought clippy itself already depended on it

How fast is the test serially? If we're only saving milliseconds we can just drop it

Reply:
`dev update_lints` will give a rough estimate of how fast it should be, so basically instant.

Reply:
Yeah, around 50ms with `dev` profile on my laptop, I removed parallelism.